xcube/core/byoa/config.py

```diff
@@ -0,0 +1,537 @@
+        :param gh_username: Optional GitHub user name.
+        :param gh_token: Optional GitHub user name.
+        :return:
+        """
+        assert_given(gh_org, 'gh_org')
+        assert_given(gh_org, 'gh_repo')
+        assert_given(gh_org, 'gh_tag')
+        assert_given(gh_org, 'gh_release')
+        gh_url = f'https://github.com/{gh_org}/{gh_repo}/archive/{gh_tag}.zip'
+        gh_sub_path = f'{gh_repo}-{gh_release}'
+        gh_params = None
+        if gh_token is not None:
+            gh_params = dict(token=gh_token, username=gh_username or gh_org)
+        elif gh_username is not None:
+            assert_given(gh_token, 'gh_token')  # fails always
+        return CodeConfig(file_set=FileSet(gh_url,
+                                           sub_path=gh_sub_path,
+                                           storage_params=gh_params),
+                          callable_ref=callable_ref,
+                          callable_params=callable_params)
+
+    def for_service(self) -> 'CodeConfig':
+        """
+        Convert this code configuration so can be used by the
+        generator service.
+
+        The returned code configuration defines either
+        inline code or a file set that represents a local
+        ZIP archive.
+        """
+        return _for_service(self)
+
+    def for_local(self) -> 'CodeConfig':
+        """
+        Convert this code configuration so can be used by the
+        local generator. This means, the returned configuration
+        can be used to load executable code. i.e.
+        :meth:get_callable() will return a callable.
+
+        The returned code configuration defines either
+        a callable (member _callable) or a file set that represents
+        a local directory. This directory may be used without
+        package installation (i.e. added to Python's sys.path) or
+        used after installation (i.e. by executing
+        ```python setup.py install```).
+        """
+        return _for_local(self)
+
+    def get_callable(self) -> Callable:
+        """
+        Get the callable specified by this configuration.
+
+        In the common case, this will require importing the callable.
+
+        :return: A callable
+        :raise ImportError if the callable can not be imported
+        """
+        if self._callable is None:
+            self.set_callable(self._load_callable())
+        return self._callable
+
+    def set_callable(self, func_or_class: Callable):
+        """
+        Set the callable that is represented by this configuration.
+        :param func_or_class: A callable
+        """
+        assert_true(callable(func_or_class), f'func_or_class must be callable')
+        self._callable = func_or_class
+
+    def _load_callable(self) -> Callable:
+        """
+        Load the callable specified by this configuration.
+
+        :return: A callable
+        :raise ImportError if the callable can not be imported
+        """
+        code_config = self.for_local()
```

Comment:
This is the part that I had the biggest problems with wrapping my head around. At a first glance, it looks like a clear error, as _load_callable is called in any case, whether it is running on a service or locally. In the documentation of 'for_local', it is even mentioned that the method shall be used to work with a local generator. However, the steps seem to be important for any config, without regard to where it is running.
It would make things clearer to have a method that is used by both 'load_callable' and 'for_local' that does not use the term 'local'.

Reply:
I guess, this is because you are confused by the meaning of _service_ here. Also the service executes _local_ code. But from the configuration perspective both the generator service and local generator are _clients_ which submit a request containing code configfuration. The distinction here is whether I create a request to be posted to a service or to the local code execution context.